### PR TITLE
Country taxonomy

### PR DESCRIPTION
**Pull request by Subsurface-divelog**
### Describe the pull request:

- [ ] Bug fix
- [x] Functional change
- [x] New feature
- [ ] Documentation change
- [ ] Language translation

Kind of none-of-the-above...
@tcanabrava implemented the dive_site country data... but the additional country field in the data structure and file formats seemed redundant with the taxonomy; so I changed things over
### Pull request long description:

Use the existing taxonomy infrastructure to hold the country information - and by default populate the field from that taxonomy, if it exists.
In the process of doing so, I discovered a few bugs in our logic of how we handled dive site edits and tried to fix them
### Changes made:

1) new helpers

2) remove country member of struct dive_site

3) remove country attribute in file formats

4) change the way we call the reverse geo tagging function

### Related issues:

See #144 
### Mentions:

@tcanabrava and others, please review and comment

### CODE REVIEW

**Comment by dirkhh** [comment]
@tcanabrava would love a review
@neolit123 - I know this isn't your core area, but your reviews tend to be thorough, so I'd appreciate if you took a look as well!

**Comment by tcanabrava** [comment]
Seems correct. +1.

File: core/divesitehelpers.cpp [hunk 1]
```diff
@@ -53,158 +43,134 @@ void ReverseGeoLookupThread::run() {
...
-			}
-			QJsonObject obj = jsonDoc.object();
-			QVariant geoNamesObject = obj.value("geonames").toVariant();
-			QVariantList geoNames = geoNamesObject.toList();
-			if (geoNames.count() > 0) {
-				QVariantMap firstData = geoNames.at(0).toMap();
-				int ri = 0, l3 = -1, lt = -1;
-				if (ds->taxonomy.category == NULL) {
-					ds->taxonomy.category = alloc_taxonomy();
+	if(timer.isActive()) {
```
Comment (neolit123) [inline core/divesitehelpers.cpp]: space after `if`.

File: core/divesitehelpers.cpp [hunk 2]
```diff
@@ -53,158 +43,134 @@ void ReverseGeoLookupThread::run() {
...
+		timer.stop();
+		if(reply->error() > 0) {
```
Comment (neolit123) [inline core/divesitehelpers.cpp]: space after `if`.

File: core/taxonomy.c [hunk 3]
```diff
@@ -47,3 +48,31 @@ int taxonomy_index_for_category(struct taxonomy_data *t, enum taxonomy_category
 			return i;
 	return -1;
 }
+
+const char *taxonomy_get_country(struct taxonomy_data *t)
+{
+	for (int i = 0; i < t->nr; i++)
+		if (t->category[i].category == TC_COUNTRY)
+			return t->category[i].value;
```
Comment (neolit123) [inline core/taxonomy.c]: [1] i was wondering about this before. in that case do we want to wrap the `for` loop with `{ }`, given it wraps more than a single line or is the above just fine.

File: core/taxonomy.c [hunk 4]
```diff
@@ -47,3 +48,31 @@ int taxonomy_index_for_category(struct taxonomy_data *t, enum taxonomy_category
...
+}
+
+void taxonomy_set_country(struct taxonomy_data *t, const char *country, enum taxonomy_origin origin)
+{
+	int idx = -1;
+	for (int i = 0; i < t->nr; i++)
+		if (t->category[i].category == TC_COUNTRY) {
+			idx = i;
+			break;
+		}
```
Comment (neolit123) [inline core/taxonomy.c]: same as [1].

**Review by neolit123 — COMMENTED** [review]
i've added a couple of small comments + there is a typo in `a6fbeedea` (`otherone`).

File: core/taxonomy.c (hunk 3 above)
Comment (dirkhh) [reply core/taxonomy.c]: From a C/C++ syntax the code as is works - but adding { } wouldn't hurt. We do have this pattern quite a few times in the code, though.

File: core/taxonomy.c (hunk 4 above)
Comment (dirkhh) [reply core/taxonomy.c]: I'll add the outer { }

File: core/divesitehelpers.cpp (hunk 1 above)
Comment (dirkhh) [reply core/divesitehelpers.cpp]: This is a pure re-indent. I want `git show -w` to be empty so it's clear the commit only deals with whitespace.
I can add another commit that fixes this.

**Comment by dirkhh** [comment]
Also fixed the typo in the commit message. Planning to wait for Travis to finish (nothing should have changed - and none of this code is even exercised, but at least it's a build check) and then merge this PR.
Thanks for the reviews!

**Comment by dirkhh** [comment]
Oh, also snuck in one additional commit (948f7b1) when updating the PR: the satellite icon wasn't in the qrc file and depending on how you build / install Subsurface, it wouldn't show up in the divesite edit dialog.

**Comment by dirkhh** [comment]
Arg, silly me. CI will fail with Kirigami.
Oh well, I'll just pull into master.
What could possibly go wrong :-)

**Comment by sfuchs79** [comment]
Unfortunately this country taxonomy thing causes now a crash on my side as soon as I try to enter a country via the "edit dive site" function for any of my dives:

[crash_entering_country.txt](https://github.com/Subsurface-divelog/subsurface/files/1360922/crash_entering_country.txt)

**Comment by dirkhh** [comment]
And annoyingly, after 10 seconds of looking at the code I know why... Thanks for the report. I'll fix that.

**Comment by sfuchs79** [comment]
With the fix no crash any more but the country I enter for a dive (any dive) is still not saved.
And I see this strange effect:
- I edit a dive A with dive site name
- I type a country name
- Apply changes
--> Country not saved for dive A
- Select a dive B w/o dive site name
- Edit dive site
--> Name and country of dive A show up in the name and country fields
- Apply changes
--> None of both is saved also for dive B

**Comment by Subsurface-divelog** [comment]
That's, err, frustrating. This (obviously, I guess) worked for me in a few combinations that I tried. But as always I guess I didn't try the combinations that you tried... but of course I can now reproduce what you are reporting here. So let's try and fix that...

**Comment by sfuchs79** [comment]
Less frustrating compared to not being able to reproduce s.th. at all :-)

And hopefully also less frustrating compared to having too less testing/testers at all. A situation like "You created a feature - nobody cares..." would be worse ;-)

Have to go to work now. Will be able to continue testing this afternoon German time.

**Comment by dirkhh** [comment]
Nur um Misverstaendnisse zu vermeiden. Ich bin SUPER DANKBAR, dass Du das alles testest und dass Du diese Probleme findest. "frustrating" da ich dachte, ich hatte alle Kombinationen getestet. Sprich, frustriert mit MIR, nicht mit Dir.

Und natuerlich (zur Belohnung) ist das ganze kompliziert da der Code einige Annahmen macht, die Du hier verletzt. Eine davon werde ich erzwingen (Du kannst keine dive site bearbeiten, die keinen Namen hat - das sollte das zweite Problem etwas einfacher machen)...

Egal, ich werde einen PR schicken und Dich drauf setzen, sobald ich das rausgeknobelt habe.

Danke

**Pull request merged**